### PR TITLE
systemd: Allow configuration of additional NTP servers.

### PR DESCRIPTION
**Pull request by mvollmer**
Note, when manually testing this, make sure that "timedatex" is not installed or when it is installed, that is is disabled (systemctl disable timedatex).  If it is enabled, the UI will not allow configuration of NTP servers.

### CODE REVIEW

**Comment by mvollmer** [comment]
f22 failure seems to be a new but unrelated one.  r7 failure is related.

**Comment by stefwalter** [comment]
@andreasn There's no label to indicate what the user is typing is a Server. Shouldn't there be a label:

"Time Servers"

File: pkg/systemd/host.css [hunk 1]
```diff
@@ -0,0 +1,31 @@
+.systime-inline {
```
Comment (stefwalter) [inline pkg/systemd/host.css]: What is this used for?

File: pkg/systemd/host.js [hunk 2]
```diff
@@ -663,9 +664,44 @@ PageSystemInformationChangeSystime.prototype = {
         $('#systime-timezones').on('change', enable_apply_button);
         $('#systime-date-input').on('focusin', $.proxy(this, "store_date"));
         $('#systime-date-input').on('focusout', $.proxy(this, "restore_date"));
+
+        this.ntp_servers_tmpl = $("#ntp-servers-tmpl").html();
```
Comment (stefwalter) [inline pkg/systemd/host.js]: Shouldn't we use 'self' here, and on the lines below? Especially in big functions, it makes the code much easier to reason about.

File: pkg/systemd/host.js [hunk 3]
```diff
@@ -663,9 +664,44 @@ PageSystemInformationChangeSystime.prototype = {
...
+        $('#systime-ntp-servers').on('click', '[data-action="add"]', function () {
+            var index = $(this).attr('data-index');
+            if (self.ntp_servers !== undefined) {
+                self.sync_ntp_servers();
+                self.ntp_servers.splice(index+1, 0, "");
+                self.update_ntp_servers();
+            }
+            // HACK - without returning 'false' here, the dialog will
+            // be magically closed when controlled by the
+            // check-system-info test.
```
Comment (stefwalter) [inline pkg/systemd/host.js]: Is this because you use 'data-action="add"'? What happens when you change it to 'data-actionx'?

File: pkg/systemd/host.js [hunk 4]
```diff
@@ -709,13 +755,107 @@ PageSystemInformationChangeSystime.prototype = {
...
+                    self.ntp_config_file = cockpit.file("/etc/systemd/timesyncd.conf.d/50-cockpit.conf",
+                                                        { superuser: "try" });
+
+                self.ntp_config_file.read().
+                    done(function (text) {
+                        if (text === null)
+                            self.ntp_servers = [ ];
+                        else {
+                            text.split("\n").forEach(function (line) {
+                                if (line.indexOf("NTP=") === 0)
```
Comment (stefwalter) [inline pkg/systemd/host.js]: I would suggest braces around a multiline if statement.

File: pkg/systemd/host.js [hunk 5]
```diff
@@ -709,13 +755,107 @@ PageSystemInformationChangeSystime.prototype = {
...
+                            });
+                        }
+                        if (self.ntp_servers.length === 0)
+                            self.ntp_servers = [ "" ];
+                        self.update_ntp_servers();
+                        self.update_ntp_status();
+                        self.update();
+                    }).
+                    fail(function (error) {
+                        console.log("error", error);
```
Comment (stefwalter) [inline pkg/systemd/host.js]: How about "failed to load time servers"?

File: pkg/systemd/host.js [hunk 6]
```diff
@@ -709,13 +755,107 @@ PageSystemInformationChangeSystime.prototype = {
...
+        if (self.ntp_servers !== undefined) {
+            self.sync_ntp_servers();
+            var servers, text;
+            var promise;
+
+            servers = self.ntp_servers.filter(function (val) { return val !== ""; });
+            if (servers.length === 0)
+                text = null;
+            else
+                text = cockpit.format("[Time]\nNTP=$0\n", servers.join(" "));
```
Comment (stefwalter) [inline pkg/systemd/host.js]: Could we add a "# This file is automatically generated by Cockpit\n" ?

File: pkg/systemd/host.js [hunk 7]
```diff
@@ -709,13 +755,107 @@ PageSystemInformationChangeSystime.prototype = {
...
+
+            return cockpit.spawn([ "mkdir", "-p", "/etc/systemd/timesyncd.conf.d" ], { superuser: "try" }).
```
Comment (stefwalter) [inline pkg/systemd/host.js]: Should we have an option in file.replace() that does this? I see this starting to spread around in the code.

**Comment by stefwalter** [comment]
Nice work.

This also needs documentation in the guide under *Feature Internals* : *systemd*  describing that we only support changing time servers with systemd-timesyncd

**Comment by andreasn** [comment]
The entry field is a bit short. Maybe we need to force a width on it.
screenshot from 2015-10-08 12-57-56

**Comment by andreasn** [comment]
>  There's no label to indicate what the user is typing is a Server.

How about something like this
screenshot from 2015-10-08 13-26-33

**Comment by andreasn** [comment]
the has-error div seems to be visible even though there is no error

**Comment by andreasn** [comment]
The "Using Time Server 79.136.86.176:123 (0.fedora.pool.ntp.org)." hangs a bit in the air, and since it doesn't have anything to do with the time zone, it would be good to connect it with the time servers controls a bit more.
screenshot from 2015-10-08 13-45-38

**Comment by andreasn** [comment]
The ip address and the human readable address are actually the same, but @mvollmer said on IRC that this is the string we get from systemd, so not much we can do here.

File: pkg/systemd/host.css (hunk 1 above)
Comment (mvollmer) [inline pkg/systemd/host.css]: This is for the list of ntp servers, #ntp-servers.  I copied it from the docker port config markup, which is very similar.

File: pkg/systemd/host.js (hunk 2 above)
Comment (mvollmer) [inline pkg/systemd/host.js]: Yes, fixed.

File: pkg/systemd/host.js (hunk 4 above)
Comment (mvollmer) [inline pkg/systemd/host.js]: Yes, fixed.

File: pkg/systemd/host.js (hunk 5 above)
Comment (mvollmer) [inline pkg/systemd/host.js]: Yes, much better.

File: pkg/systemd/host.js (hunk 6 above)
Comment (mvollmer) [inline pkg/systemd/host.js]: Yes, done.

File: pkg/systemd/host.js (hunk 7 above)
Comment (mvollmer) [inline pkg/systemd/host.js]: I think so.  In the bridge or in cockpit.js?

File: pkg/systemd/host.css (hunk 1 above)
Comment (stefwalter) [inline pkg/systemd/host.css]: We shouldn't need to define empty classes.

File: pkg/systemd/host.js (hunk 7 above)
Comment (stefwalter) [inline pkg/systemd/host.js]: I think in the bridge would be appropriate. An option on the fsreplace channel.

**Comment by mvollmer** [comment]
> the has-error div seems to be visible even though there is no error

There actually never is a error, so I have removed that table row,

**Comment by mvollmer** [comment]
> This also needs documentation in the guide under *Feature Internals* : *systemd*  describing that we only support changing time servers with systemd-timesyncd

Done.

**Comment by mvollmer** [comment]
> How about something like this using the placeholder tag.

Done.  The placeholder is only for the first entry, right?

**Comment by mvollmer** [comment]
>  it would be good to connect it with the time servers controls a bit more.

Done.

File: pkg/systemd/host.css (hunk 1 above)
Comment (mvollmer) [inline pkg/systemd/host.css]: Ok, now I get it, sorry.  Fixed.

File: pkg/systemd/host.js (hunk 7 above)
Comment (mvollmer) [inline pkg/systemd/host.js]: Ok, should we do this here or in another PR?

File: pkg/systemd/host.js (hunk 3 above)
Comment (mvollmer) [inline pkg/systemd/host.js]: I changed that to data-action-add and data-action-del, without any effect.

**Comment by andreasn** [comment]
> Done. The placeholder is only for the first entry, right?

I think so, but I'm not super-sure.
It could be the case that another sysadmin logged in and set time servers six months ago. Now another sysadmin needs to change it today. That the custom entry is a NTP server might therefore not be obvious.
Probably doesn't hurt to have it in all entrys

**Comment by stefwalter** [comment]
When I select 'Automatically using NTP' the UI feels like it's forcing me to type an NTP server. This is not at all the case though.

Should there be three options for 'Set Time'?

 * Manually
 * Automatically using NTP
 * Automatically using specific NTP servers

**Comment by stefwalter** [comment]
In addition there is an [x] button which does nothing when i click it on the first row.

**Comment by stefwalter** [comment]
Having this as three options in the dropdown also cleanly isolates the mode where we can set specific NTP servers. We can disable that option when timedatex is installed.

**Comment by mvollmer** [comment]
> In addition there is an [x] button which does nothing when i click it on the first row.

It will remove that row, but if it is the only row and it is empty, there will be no change since a new, empty row will immediately appear.

**Comment by mvollmer** [comment]
> Should there be three options for 'Set Time'?

I like this.  This avoids the current confusion that not specifying any time servers actually means "use the default".  We should make it so that selecting "Automatically using specific NTP servers" forces you to specify at least one time server.

**Comment by andreasn** [comment]
> Should there be three options for 'Set Time'?

That's not a bad idea at all. Would make it clearer what is being used, and highlights that the default servers are not removed, just overridden.

**Comment by andreasn** [comment]
Mockup

**Comment by stefwalter** [comment]
Looks much clearer. Nice.

**Comment by mvollmer** [comment]
> Should there be three options for 'Set Time'?

Done.

**Comment by stefwalter** [comment]
I still see only two options for 'Set Time'. Am i missing something?

**Comment by mvollmer** [comment]
> I still see only two options for 'Set Time'. Am i missing something?

There is nothing missing in this PR, as far as I can tell.  Could you double-check whether you are testing the right bits?

**Comment by mvollmer** [comment]
So I now think that the NTP status really doesn't belong into the dialog.  I'll put it on the page, trusting that we find a nice layout for it later.

**Comment by mvollmer** [comment]
>  I'll put it on the page, trusting that we find a nice layout for it later.

Done.

**Comment by mvollmer** [comment]
@andreasn, ready for your review.

**Comment by mvollmer** [comment]
Now the status is displayed via a help bubble thing.  I couldn't quickly get the popovers working, so I made it show on hover.

I can work more on this later today.

**Comment by mvollmer** [comment]
I realized that a status like "Using time server ..." from timesyncd does not mean that it is actually getting any useable time from it.  We also need to show the NTPSynchronized property of timedate1.

**Comment by andreasn** [comment]
I didn't get the (i) icon until I switched to a custom server. I switched back to the default and it remained.
Did it fail to show up the first time?

**Comment by andreasn** [comment]
> I didn't get the (i) icon until I switched to a custom server. I switched back to the default and it remained.

I tried to reproduce this with reboots and stuff, but wasn't able to. So either it was a temporary glitz or some clipping happening due to me having the inspector open.

**Comment by andreasn** [comment]
Fixed input widths here: https://github.com/andreasn/cockpit/tree/systemd-timesync-inputs

**Comment by andreasn** [comment]
How about having the error be "Failed to synchronize with server example.com" to make it more descriptive?

**Comment by mvollmer** [comment]
> How about having the error be "Failed to synchronize with server example.com" to make it more descriptive?

In the general case, we would have to list all servers:

    Failed to synchronize with any of 0.pool.ntp.org, 1.pool.ntp.org, and 2.pool.ntp.org

    Failed to synchronize with any of the default servers. 

I think this isn't necessary, or?

We can link to the timesyncd service page which has the log of all recent failures, though. What about that?

Super fance would be to include in the popover the errors from the log that have happened after the most recent configuration change...

**Comment by mvollmer** [comment]
> Fixed input widths here: https://github.com/andreasn/cockpit/tree/systemd-timesync-inputs

Included here.

**Comment by andreasn** [comment]
> We can link to the timesyncd service page which has the log of all recent failures, though. What about that?

That's a nice idea!

**Comment by mvollmer** [comment]
DBusClient.poll etc need docs.

File: pkg/base1/cockpit.js [hunk 8]
```diff
@@ -1853,6 +1854,29 @@ function full_scope(cockpit, $, po) {
             return new DBusProxies(self, cache, String(iface), String(path_namespace));
         };
 
+        self.poll = function poll(path, iface) {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: I think this should be implemented as a call through to cockpit_dbus_cache_poke() instead. We could have a dbus JSON "poke" message. The actual property changes would then flow back up through the usual "notify" channels.

Comment (mvollmer) [inline pkg/base1/cockpit.js]: Maybe the bridge can be instructed to do the polling so that there isn't any traffic unless there is a actual change.  It already 'fixes up' properties that are merely invalidated (instead of being updated), it could also optionally fix up properties that never receive notifications.

Comment (stefwalter) [inline pkg/base1/cockpit.js]: True, it could be an optional setting. Perhaps an argument to the "watch" message? Because we already 'poke' whenever there is an outgoing message.

Comment (mvollmer) [inline pkg/base1/cockpit.js]: I'll factor this out of this PR.

Comment (mvollmer) [inline pkg/base1/cockpit.js]: Here: https://github.com/cockpit-project/cockpit/issues/2970

**Comment by dperpeet** [comment]
needs rebase

**Comment by stefwalter** [comment]
Unfortunately this happens on rhel-7:

```
ln -s '/usr/lib/systemd/system/test.service' '/etc/systemd/system/default.target.wants/test.service'
> can't get NTPSynchronized property Unknown property
> can't get NTPSynchronized property Unknown property
> can't get NTPSynchronized property Unknown property
Unexpected journal message 'org.freedesktop.systemd1: couldn't get property org.freedesktop.systemd1.Service ExecMain at /org/freedesktop/systemd1/unit/systemd_2dtimedated_2eservice: GDBus.Error:org.freedesktop.DBus.Error.UnknownProperty: Unknown property'
Unexpected journal message 'org.freedesktop.systemd1: couldn't get property org.freedesktop.systemd1.Service ExecMain at /org/freedesktop/systemd1/unit/systemd_2dtimedated_2eservice: GDBus.Error:org.freedesktop.DBus.Error.UnknownProperty: Unknown property'
Warning: Permanently added '10.111.121.183' (ECDSA) to the list of known hosts.
Journal database copied to TestServices-testBasic-10.111.121.183-FAIL.journal
not ok 5 testBasic (check_services.TestServices)
Traceback (most recent call last):
  File "/home/hubbot/hubbot/7f8797ab495e10dcfbec29593d4c148d19e4d11f_r7_x86-64/test/testlib.py", line 480, in tearDown
    self.check_journal_messages()
  File "/home/hubbot/hubbot/7f8797ab495e10dcfbec29593d4c148d19e4d11f_r7_x86-64/test/testlib.py", line 611, in check_journal_messages
    raise Error("There were unexpected journal messages")
Error: There were unexpected journal messages
```

The f22 failure is #2978

**Comment by mvollmer** [comment]
> `> can't get NTPSynchronized property Unknown property`

I'll suppress that message.

The real failure comes from this:

```
Unexpected journal message 'org.freedesktop.systemd1: couldn't get property org.freedesktop.systemd1.Service ExecMain at /org/freedesktop/systemd1/unit/systemd_2dtimedated_2eservice: GDBus.Error:org.freedesktop.DBus.Error.UnknownProperty: Unknown property'
```

This seems to happen when the bridge receives a PropertyNotify signl and then goes to retrieve the values for the invalidated properties, at which point systemd pretends that the property is unknown.

Should we filter that out in general?

**Comment by stefwalter** [comment]
> Should we filter that out in general?

Yes, it's an race that's bound to present when the DBus daemon doesn't support ObjectManager ... and is in fact in part why ObjectManager was built. So lets turn that into a g_debug().

**Comment by mvollmer** [comment]
> So lets turn that into a g_debug().

Ok, I'll do that here.

**Comment by mvollmer** [comment]
> Could you use dbus_error_matches_unknown() and only use g_debug in that case?

Yes, very good.
